frontend/projects/upgrade/src/app/core/experiment-design-stepper/experiment-design-stepper.service.ts

```diff
@@ -294,45 +302,76 @@ export class ExperimentDesignStepperService {
-
-    const factorOne = designData.factors[0];
-    const factorTwo = designData.factors[1];
-
-    factorOne.levels.forEach((factorOneLevel) => {
-      factorTwo.levels.forEach((factorTwoLevel) => {
-        const tableRow: FactorialConditionTableRowData = {
-          id: uuidv4(), // TODO: maybe not the right place?
-          levels: [
-            {
-              id: factorOneLevel.id,
-              name: factorOneLevel.name,
-            },
-            {
-              id: factorTwoLevel.id,
-              name: factorTwoLevel.name,
-            },
-          ],
-          condition: this.createFactorialAliasString(
-            factorOne.name,
-            factorOneLevel.name,
-            factorTwo.name,
-            factorTwoLevel.name
-          ),
-          alias: '',
-          weight: '0.0',
-          include: true,
-        };
+    requiredFactorialTableData.map((conditionData) => {
+      const conditionLevelsData = this.filterLevelsData(conditionData);
+      const conditions = this.createConditionString(conditionData);
 
-        tableData.push(tableRow);
-      });
+      const tableRow: FactorialConditionTableRowData = {
+        id: uuidv4(), // TODO: maybe not the right place?
+        levels: conditionLevelsData,
+        condition: conditions,
+        alias: '',
+        weight: '0.0',
+        include: true,
+      };
+      console.log("tableRow.condition.split(';').length");
+      console.log(tableRow.condition.split(';').length);
+      tableData.push(tableRow);
     });
 
+    // console.log('tableData');
+    // console.log(tableData);
     return tableData;
   }
 
+  factorDataToConditions(factorsData: ExperimentFactorFormData[], levelsCombinationData: FactorLevelData[] = []) {
+    // return if no data in factors
+    if (factorsData.length === 0) {
+      return [levelsCombinationData];
+    } else {
+      // taking the 1st factor
+      const currentFactor = factorsData[0];
+      const levelPermutations = [];
+
+      for (let i = 0; i < currentFactor.levels.length; i++) {
+        const levelId = currentFactor.levels[i].id;
+        const levelName = currentFactor.levels[i].name;
+
+        // taking level of current factor and processing on other factors
+        const remainingLevelsPermutations = this.factorDataToConditions(factorsData.slice(1), [
+          ...levelsCombinationData,
+          { factor: currentFactor.name, id: levelId, name: levelName },
+        ]);
+        levelPermutations.push(...remainingLevelsPermutations);
+      }
+      return levelPermutations;
+    }
+  }
+
+  filterLevelsData(conditionData: FactorLevelData[]) {
+    const levels: FactorialLevelTableRowData[] = [];
+    conditionData.forEach((level) => {
+      levels.push({ id: level.id, name: level.name });
+    });
+    return levels;
+  }
+
+  createConditionString(conditionData: FactorLevelData[]) {
+    let alias: string;
+    let count = 0;
+    conditionData.forEach((level) => {
```

Comment:
Couple nitpicky things, ignore me if you like. fyi `Array.forEach` comes with an index argument: `conditionData.forEach((level, index)`, you don't have to track it manually with `count`.

Also I don't know if there's a possibility if this, but I notice that if `conditionData` is empty somehow, this function will return `undefined` for alias, which I doubt is a happy result. Maybe just initialize as empty string?

Reply:
resolved both